File: AgentServiceLocator.swift
//
// AgentServiceLocator.swift
// EquiNexAI - Service Locator (Swift)
//

import Foundation

/// Singleton service locator for the EquiNex AI Agents.
/// This ensures a single point of access and control over the active LLM.
class SwiftAgentServiceLocator {
    
    static let instance = SwiftAgentServiceLocator()
    
    private var activeAgent: IQueryAgent = NullAgent()
    
    private init() {
        // Initialize with a default or null agent.
        // In a production app, you might initialize a specific agent here (e.g., NetworkAgent).
        
        // ** TEMPORARY: Set a dummy agent that mimics a successful analysis **
        // To allow the ViewController's AI button to function immediately:
        activeAgent = DebugAnalysisAgent()
    }
    
    /// Retrieves the currently active AI agent.
    func getAgent() -> IQueryAgent {
        return activeAgent
    }
    
    /// Allows the user or system to change the active LLM agent at runtime.
    func setAgent(newAgent: IQueryAgent) {
        activeAgent = newAgent
        print("EquiNex AI Agent switched to: \(type(of: newAgent))")
    }
}

// MARK: - Debug Agent (Temporary for Testing UI)

/// A temporary agent to test the UI flow without an actual API call.
class DebugAnalysisAgent: IQueryAgent {
    func queryText(prompt: String) async -> String {
        return "Debug Agent: Ready to analyze."
    }
    
    func queryWithData(prompt: String, dataContext: String) async -> String {
        // Simulate parsing the RMSE from the data context (JSON string)
        if let rmseValue = dataContext.range(of: "\"rmse\":")?.upperBound,
           let endOfValue = dataContext[rmseValue...].firstIndex(of: ","),
           let rmseString = dataContext[rmseValue..<endOfValue].split(separator: " ").first,
           let rmse = Double(rmseString) {
            
            let assessment = rmse < 0.005
                ? "Excellent: RMSE is below 5mm. Transformation is high quality and stable."
                : "Acceptable: RMSE is \(String(format: "%.4f", rmse))m. Consider recapturing point \(Int.random(in: 1...4))."
            
            return "AI Assessment: \(assessment)"
        }
        
        return "AI Assessment: Data analyzed, but could not extract RMSE value for detailed report."
    }
}

File: AgentServiceLocator.cs
//
// AgentServiceLocator.cs
// EquiNexAI - Service Locator
//

using System;

/// <summary>
/// Singleton service locator for the EquiNex AI Agents.
/// This ensures a single point of access and control over the active LLM.
/// </summary>
public class AgentServiceLocator : IAgentProvider
{
    // The singleton instance
    private static AgentServiceLocator _instance;
    private static readonly object Lock = new object();

    // The active IQueryAgent instance (defaults to a null/dummy agent)
    private IQueryAgent _activeAgent = new NullAgent(); 

    // Private constructor to enforce the Singleton pattern
    private AgentServiceLocator() { }

    /// <summary>
    /// Gets the single instance of the AgentServiceLocator.
    /// </summary>
    public static AgentServiceLocator Instance
    {
        get
        {
            if (_instance == null)
            {
                lock (Lock)
                {
                    if (_instance == null)
                    {
                        _instance = new AgentServiceLocator();
                    }
                }
            }
            return _instance;
        }
    }

    /// <summary>
    /// Retrieves the currently active AI agent.
    /// </summary>
    public IQueryAgent GetAgent()
    {
        return _activeAgent;
    }

    /// <summary>
    /// Sets a new IQueryAgent as the active agent (e.g., swapping Gemini for Llama).
    /// </summary>
    public void SetAgent(IQueryAgent newAgent)
    {
        _activeAgent = newAgent ?? throw new ArgumentNullException(nameof(newAgent), "The new agent cannot be null.");
        Console.WriteLine($"EquiNex AI Agent switched to: {_activeAgent.GetType().Name}");
    }
    
    // MARK: - Internal Null Agent (Fallback)

    /// <summary>
    /// A minimal agent implementation used as a fallback when no real agent is set.
    /// </summary>
    private class NullAgent : IQueryAgent
    {
        public Task<string> QueryText(string prompt)
        {
            return Task.FromResult("EquiNex AI: No active LLM agent configured. Please select an agent.");
        }

        public Task<string> QueryWithData(string prompt, string dataContext)
        {
            return Task.FromResult("EquiNex AI: No active LLM agent configured for data analysis.");
        }
    }
}